### ISSUE TITLE
how to use the training results

### ISSUE THREAD

**Issue by eejlny** [Xilinx/pytorch-ocr#2]
Hello, 

I would appreciate very much if you could give some further insights on how to take the output of the training process into the pynq hardware. 

Basically I have the LSTM running correctly in a Zynq Board with the design files provided in https://github.com/Xilinx/LSTM-PYNQ. 

Now I have used the setup in here to retrain the system for a different problem. In the checkpoints directory there are the tar files with the best word error rate.  I would like to use these training results in the board but I cannot find some instructions on how to do that. Could you advise, please ?

THanks,

**Comment by volcacius** [comment]
Hi, 

As the README says, you can invoke the main with the --export flag together with the desired number of SIMD and PE (--simd_factor and --pe), together with a pretrained model (--input). If you look at how the function call propagates through the code you can get a picture of how _r_model_fw_bw.h_ is generated. 

Alessandro

**Comment by eejlny** [comment]
Hello,


Thanks for the answer.


I have been trying to do that for several days but it raises exceptions so when it runs.


I have uploaded the .tar and .json inputs and command here in a zip file here:


https://github.com/Xilinx/LSTM-PYNQ/issues/6


I have tried with and without .json and with and without specifying --pe and --simd_factor.


Could you try with these inputs to see if it runs for you ? I am not sure if the problem is in the inputs or somewhere else.


Thanks,


Reader in Adaptive and Energy Efficient Computing
Fellow of the Higher Education Academy
Advanced Microelectronics and System Engineering (AMSE) MSc program director
Merchant Venturers Building
Department of Electronic and Electrical engineering
University of Bristol
phone : 0117 3315128
Fax : 0117 9545205
web: http://seis.bris.ac.uk/~eejlny/
e_mail : j.l.nunez-yanez@bristol.ac.uk<mailto:j.l.nunez-yanez@bristol.ac.uk>
________________________________

**Comment by eejlny** [comment]
Hello,


OK, so I got the script to run to completion but removing the 32 bit precision in  the example:


quantized_params/QLSTM128_W2B8A4I32_FC_W32B32.json


for example like:


quantized_params/QLSTM128_W2B8A4I8_FC_W8B8.json


I guess what I am not understanding is the part that says:

### Strategy

Training a model that can be reproduced accurately in hardware requires at least one retraining step. The reason is that the batch norm coefficients have to be fused into the quantized fully connected layer, since batch norm is not implemented in hardware.
The suggested strategy goes as follow:

* Train with quantized input, quantized recurrent weights/activations/bias, full-precision internal activations and full-precision fully connected layer, either from scratch or from a pretrained full precision model.
* Perform batch norm-fully connected fusion and retrain with quantized internal activations and quantized fully connected layer.


I am just training with QLSTM128_W2B8A4I32_FC_W32B32.json with

python main.py -p quantized_params/QLSTM128_W2B8A4I32_FC_W32B32.json

and then trying to obtain the hardware hls file with --export.

Could you clarify what I need to do for
* Perform batch norm-fully connected fusion and retrain with quantized internal activations and quantized fully connected layer.

with an example, please?

Thanks,


Reader in Adaptive and Energy Efficient Computing
Fellow of the Higher Education Academy
Advanced Microelectronics and System Engineering (AMSE) MSc program director
Merchant Venturers Building
Department of Electronic and Electrical engineering
University of Bristol
phone : 0117 3315128
Fax : 0117 9545205
web: http://seis.bris.ac.uk/~eejlny/
e_mail : j.l.nunez-yanez@bristol.ac.uk<mailto:j.l.nunez-yanez@bristol.ac.uk>
________________________________
